sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala

```diff
@@ -322,6 +322,27 @@ trait BatchWriteHelper {
   }
 }
 
+trait V2ExistingTableWriteExec extends V2TableWriteExec {
+  def session: SparkSession
+  def relation: DataSourceV2Relation
+  def write: Option[BatchWrite] = None
```

Comment:
I kept it as `Option` just in case we don't want to apply the new logic all the time and will introduce a flag to fallback to the old approach. If we are not going to have that flag, we can make this required.

Reply:
Cool, it would simplify the logic if we know the new code will always be applied. Perhaps worth creating a JIRA to track this.